source/language/types.rst

```diff
@@ -359,3 +410,94 @@ variables whose values may only be known at run time.
    let last_three = two[-4:-1];
    // Concatenate two alias in another one
    let both = sliced || last_three;
+
+Classical value bit slicing
+---------------------------
+
+A subset of classical values (int, uint, and angle) may be accessed at the bit
+level using index sets similar to register slicing. The bit slicing operation
+always returns a bit array of size equal to the size of the index set.
+
+.. code-block:: c
+
+   int[32] myInt = 15; // 0xF or 0b1111
+   bit[1] lastBit = myInt[0]; // 1
+   bit[1] signBit = myInt[31]; // 0
+   bit[1] alsoSignBit = myInt[-1]; // 0
+
+   bit[16] evenBits = myInt[0:2:31]; // 3
+   bit[16] upperBits = myInt[-16:-1];
+   bit[16] upperReversed = myInt[-1:-16];
+
+   myInt[4:7] = "1010"; // myInt == 0xAF
+
+For the purpose of accessing bit-level portions of array element values, a
+special ``#`` signifier may optionally be inserted between the array accessor
+and the bit accessor. This can help with readability and the compiler's ability
+to provide good error messages.
+
+.. code-block:: c
+
+   array[int[32], 5] intArr = {0, 1, 2, 3, 4};
+   // Access bit 0 of element 0 of intArr and set it to 1
+   intArr[0]#[0] = 1;
+   intArr[1][0] = 0; // The '#' is optional
+   // lowest 5 bits of intArr[4] copied to b
+   bit[5] b = intArr[4]#[0:4];
+
+.. _array-slicing:
+
+Array concatenation and slicing
+-------------------------------
+
+Two or more classical arrays can be
+concatenated to form an array of the same type whose size is the
+sum of the sizes of the individual arrays. Unlike with qubit registers, this operation
+copies the contents of the input arrays to form the new (larger) array. This means that
+arrays *can* be concatenated with themselves.
```

Comment:
If we allow arrays of arbitrary length to be passed to subroutines (e.g. `def f(array[int, dim(1)] v)`), then we need to forbid the concatenation operator from being used in a subroutine parameter list.

Reply:
> we need to forbid the concatenation operator from being used in a subroutine parameter list.

This is confusing to me because I was not 100% clear on the language here until I looked it up. Are you referring to the parameter list associated with the subroutine definition, or the argument list associated with the subroutine call?

Reply:
Based on our discussion during the last WG meeting I'm assuming you mean the argument list associated with the call.

Reply:
Yeah, in the call, sorry.  I mean that
```text
def f(const array[int, ...] x) { }

f(a || b)
```
is an invalid call, because `a || b` doesn't resolve to a memory location.  My whole comment wasn't very well written, sorry.  This _can_ be forbidden grammatically, but it doesn't necessarily need to be; it can also just fail during semantic analysis.  The main point is that it's a caveat on passing arrays to function bodies.

Reply:
Added in https://github.com/Qiskit/openqasm/pull/283/commits/b05dc4bbe35ba6dee930c10e5d2f7af46860edcc